### PR TITLE
[profile] add profiler communication operations

### PR DESCRIPTION
**Pull request by 1SAA**

### CODE REVIEW

**Comment by FrankLeeeee** [comment]
Hi, I think profiler should be an independent module. You can put your code in `colossalai.utils.profile`. The memory profiler code can be put here as well @xcetos .

File: colossalai/communication/profiler.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+                  op: ReduceOp = ReduceOp.SUM,
+                  pm: ParallelMode = ParallelMode.GLOBAL,
+                  async_op: bool = False,
+                  enable_profile: bool = False) -> None:
+    comm_size = gpc.get_world_size(pm)
+
+    if comm_size == 1:
+        return
+
+    CCPROF.wait_async_op()
```
Comment (feifeibear) [inline colossalai/communication/profiler.py:147]: There are some duplicated code in different function.
Can you implement it as a wrapper?

Comment (feifeibear) [reply colossalai/communication/profiler.py:147]: https://www.geeksforgeeks.org/function-wrappers-in-python/

File: colossalai/communication/__init__.py [hunk 2]
```diff
@@ -14,4 +16,6 @@
     'send_backward_recv_backward', 'send_backward_recv_forward',
     'send_forward_recv_backward', 'recv_backward', 'recv_forward',
     'ring_forward', 'send_tensor_meta', 'recv_tensor_meta',
+    'CCPROF', 'cc_wait_aync', 'cc_all_reduce', 'cc_reduce_scatter',
+    'cc_all_gather', 'cc_broadcast', 'cc_reduce'
```
Comment (feifeibear) [inline colossalai/communication/__init__.py:20]: cc_ means colossal?

Comment (1SAA) [reply colossalai/communication/__init__.py:20]: cc = colossal communication. Anyway, it can be changed.

**Comment by feifeibear** [comment]
@FrankLeeeee Let's make a rule. The prefix of colossal-ai function is clsl_?

**Comment by kurisusnowdeng** [comment]
> @FrankLeeeee Let's make a rule. The prefix of colossal-ai function is clsl_?

Maybe col_ can be easier to understand?

**Comment by feifeibear** [comment]
> 

Of course. @1SAA use col_ as the prefix

**Comment by FrankLeeeee** [comment]
Seems @ver217  uses ca

**Comment by ver217** [comment]
> 

I used `col_` for kernel fusion functions. But for `ShardedOptimizer` and `ShardedModel`, we use `ca` now. We should unify the prefix.

**Comment by FrankLeeeee** [comment]
> > 
> 
> I used `col_` for kernel fusion functions. But for `ShardedOptimizer` and `ShardedModel`, we use `ca` now. We should unify the prefix.

Ok, let's standardize to use col_

File: colossalai/communication/profiler.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+
+class CommProf(object):
+
+    def __init__(self):
+        super().__init__()
+        self.total_count = 0
+        self.total_comm_vol = 0
+        self.total_cuda_time = 0
+        self.depth = 3
```
Comment (feifeibear) [inline colossalai/communication/profiler.py:56]: pass it through parameter instead of hard coding.

File: colossalai/communication/profiler.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+def async_check():
+    if CCPROF.pending_op is not None:
+        CCPROF.warn_flag = True
+        CCPROF.wait_async_op()
+
+
+def all_reduce(tensor: torch.Tensor,
+               op: ReduceOp = ReduceOp.SUM,
+               group=None,
+               async_op: bool = False) -> CommHandler:
```
Comment (feifeibear) [inline colossalai/communication/profiler.py:182]: return value type hint is wrong.
-> Option[CommHandler, None]

File: colossalai/communication/profiler.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        return CommHandler()
+
+    CCPROF.close_profiler(group)
+
+
+def reduce_scatter(output: torch.Tensor,
+                   input_list: List[torch.Tensor],
+                   op: ReduceOp = ReduceOp.SUM,
+                   group=None,
+                   async_op: bool = False) -> CommHandler:
```
Comment (feifeibear) [inline colossalai/communication/profiler.py:201]: ret value hint wrong.

File: colossalai/trainer/hooks/_lr_scheduler_hook.py [hunk 6]
```diff
@@ -29,6 +29,7 @@ def __init__(
         self.store_lr_in_state = store_lr_in_state
 
     def after_hook_is_attached(self, trainer):
+        self._check_metric_states_initialization(trainer)
```
Comment (feifeibear) [inline colossalai/trainer/hooks/_lr_scheduler_hook.py:32]: Is this line of code related to the profiler?

Comment (1SAA) [reply colossalai/trainer/hooks/_lr_scheduler_hook.py:32]: A small bug should be fixed when using learning rate scheduler alone.

File: colossalai/communication/__init__.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
                   recv_backward)
 from .ring import ring_forward
 from .utils import send_tensor_meta, recv_tensor_meta
+from .profiler import (CCPROF, prof_init)
```
Comment (FrankLeeeee) [inline colossalai/communication/__init__.py:9]: Can you shift this to `colossalai.utils.profile` so that memory and bandwidth profiling can be put in one module?

File: colossalai/communication/profiler.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+CCPROF = CommProf()
+torch_all_reduce = dist.all_reduce
+torch_all_gather = dist.all_gather
+torch_reduce_scatter = dist.reduce_scatter
+torch_broadcast = dist.broadcast
+torch_reduce = dist.reduce
+
+
+def prof_init(depth: int = 0):
```
Comment (FrankLeeeee) [inline colossalai/communication/profiler.py:164]: Do write the method and class name properly for clarity, e.g. `initialize_profiling`. This applies to other naming in this PR as well, i.e. `CommProf`, `CCPROF`, `CommEvent`.

File: colossalai/communication/profiler.py [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+def _get_code_location(depth: int):
+    upper_frame = inspect.stack()[depth]
+    function_name = inspect.stack()[depth - 1].function
+    info = upper_frame.filename + "(" + str(upper_frame.lineno) + "): " + function_name
+    return info
+
+
+def _format_bandwith(volme: float, time_us: int):
+    MB_PER_SECOND = (1000.0 / 1024.0) ** 2
```
[the remark this answers is not in the document]
Comment (1SAA) [reply colossalai/communication/profiler.py]: DONE

File: colossalai/communication/profiler.py (hunk 3 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: DONE

File: colossalai/communication/profiler.py (hunk 4 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: DONE

File: colossalai/communication/profiler.py (hunk 5 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: DONE

File: colossalai/communication/__init__.py (hunk 7 above)
Comment (1SAA) [reply colossalai/communication/__init__.py]: DONE

File: colossalai/communication/profiler.py (hunk 8 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: Usually, COMM = COMMUNICATION. I have added comments to point out it.

Comment (FrankLeeeee) [reply colossalai/communication/profiler.py]: If this CommXXX is not exposed to the user, I can accept this.

File: colossalai/communication/__init__.py (hunk 7 above)
Comment (FrankLeeeee) [reply colossalai/communication/__init__.py]: I mean `colossalai.utils.profile` should be a module, i.e. `colossalai.utils.profile.comm_profiler.py`, so that memory profiler can be put together as well.

**Comment by FrankLeeeee** [comment]
@1SAA Can you add unit testing code?

File: colossalai/communication/profiler.py (hunk 8 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: Users do not need to know anything about CommXXX, all they need to add is initialize_profiling and COL_COMM_PROF.show().

Comment (FrankLeeeee) [reply colossalai/communication/profiler.py]: Can we wrap `COL_COMM_PROF.show` in a function like `show_profile_results`, we should avoid letting users access the global constant.

**Comment by 1SAA** [comment]
> @1SAA Can you add unit testing code?

Done

File: colossalai/communication/profiler.py (hunk 8 above)
Comment (FrankLeeeee) [reply colossalai/communication/profiler.py]: @1SAA

File: colossalai/communication/__init__.py (hunk 7 above)
Comment (1SAA) [reply colossalai/communication/__init__.py]: Done

File: colossalai/communication/profiler.py (hunk 8 above)
Comment (1SAA) [reply colossalai/communication/profiler.py]: Done

**Review by FrankLeeeee — APPROVED** [review]
